src/bson.c

```diff
@@ -52,15 +52,162 @@ static void php_phongo_bson_visit_corrupt(const bson_iter_t* iter ARG_UNUSED, vo
+
+	php_phongo_field_path_write_item_at_current_level(state->field_path, key);
+	path_string = php_phongo_field_path_as_string(state->field_path);
+
+	phongo_throw_exception(PHONGO_ERROR_UNEXPECTED_VALUE TSRMLS_CC, "Detected unknown BSON type 0x%02hhx for field path \"%s\". Are you using the latest driver?", v_type_code, path_string);
+
+	efree(path_string);
+} /* }}} */
+
+/* Path builder */
+char* php_phongo_field_path_as_string(php_phongo_field_path* field_path)
+{
+	size_t length = 1; /* NULL character */
+	size_t i;
+	char*  path;
+	char*  ptr;
+
+	if (!field_path) {
+		return NULL;
+	}
+	if (!field_path->elements) {
+		return NULL;
+	}
+
+	for (i = 0; i <= field_path->current_level; i++) {
+		if (!field_path->elements[i]) {
+			continue;
+		}
+		length += (1 + strlen(field_path->elements[i]));
+	}
+
+	path = emalloc(length);
+	ptr  = path;
+
+	for (i = 0; i <= field_path->current_level; i++) {
+		if (!field_path->elements[i]) {
+			continue;
+		}
+		strcpy(ptr, field_path->elements[i]);
+		ptr += strlen(field_path->elements[i]);
+		ptr[0] = '.';
+		ptr++;
+	}
+	ptr[-1] = '\0';
+
+	return path;
+}
+
+php_phongo_field_path* php_phongo_field_path_alloc(void)
+{
+	php_phongo_field_path* tmp = ecalloc(1, sizeof(php_phongo_field_path));
+	tmp->ref_count             = 1;
+
+	return tmp;
+}
+
+void php_phongo_field_path_free(php_phongo_field_path* field_path)
+{
+	size_t i;
+
+	if (field_path->elements) {
+		for (i = 0; i < field_path->allocated_levels; i++) {
+			if (!field_path->elements[i]) {
+				continue;
+			}
+			efree(field_path->elements[i]);
+		}
+		efree(field_path->elements);
+	}
+	efree(field_path);
+}
+
+static void php_phongo_field_path_ensure_allocation(php_phongo_field_path* field_path, size_t level)
+{
+	if (level >= field_path->allocated_levels) {
+		size_t i;
+
+		field_path->allocated_levels = field_path->current_level + 8;
+		field_path->elements         = erealloc(field_path->elements, sizeof(char**) * field_path->allocated_levels);
+
+		for (i = level; i < field_path->allocated_levels; i++) {
+			field_path->elements[i] = NULL;
+		}
+	}
+}
+
+void php_phongo_field_path_write_item_at_current_level(php_phongo_field_path* field_path, const char* element)
+{
+	php_phongo_field_path_ensure_allocation(field_path, field_path->current_level);
+
+	if (field_path->elements[field_path->current_level] != NULL) {
+		efree(field_path->elements[field_path->current_level]);
+	}
+
+	field_path->elements[field_path->current_level] = estrdup(element);
+}
+
+bool php_phongo_field_path_push(php_phongo_field_path* field_path, const char* element)
+{
+	php_phongo_field_path_write_item_at_current_level(field_path, element);
+
+	field_path->current_level++;
+
+	return true;
+}
+
+bool php_phongo_field_path_pop(php_phongo_field_path* field_path)
+{
+	field_path->current_level--;
```

Comment:
Would it make sense for this to also check if `field_path->elements[field_path->current_level]` is null before decrementing? Currently, it looks like this might leave an element beyond the post-pop current element allocated. That runs contrary to logic you have in `php_phongo_field_path_ensure_allocation()` to ensure that pointers beyond the current level are set to null.

Reply:
In the interest of just ensuring unused pointers are set to `NULL`, let's do the follwowing before decrementing the current level, too:

```c
if (field_path->elements[field_path->current_level] != NULL) {
	field_path->elements[field_path->current_level] = NULL;
}
```